pycbc/io/hdf.py

```diff
@@ -500,54 +618,115 @@ def get_param_names(cls):
             if type(m[1]) == property]
 
     def apply_mask(self, logic_mask):
-        """Apply a boolean array to the set of triggers"""
-        if hasattr(self.mask, 'dtype') and (self.mask.dtype == 'bool'):
+        """Apply a mask over the top of the current mask
+
+        Parameters
+        ----------
+        logic_mask : boolean array or numpy array of indices
+        """
+        if self.mask is None:
+            self.mask = np.zeros(self.ntriggers, dtype=bool)
+            self.mask[logic_mask] = True
+        elif hasattr(self.mask, 'dtype') and (self.mask.dtype == 'bool'):
             orig_indices = self.mask.nonzero()[0][logic_mask]
             self.mask[:] = False
             self.mask[orig_indices] = True
         else:
             self.mask = list(np.array(self.mask)[logic_mask])
 
+    def and_masks(self, logic_mask):
+        """Apply a mask to be combined as a logical and with the current mask.
+
+        Parameters
+        ----------
+        logic_mask : boolean array or numpy array/list of indices
+        """
+        if self.mask is None:
+            # No mask exists, just update to use the given mask
+            self.apply_mask(logic_mask)
+            return
+
+        # Use intersection of the indices of True values in the masks
+        if hasattr(logic_mask, 'dtype') and (logic_mask.dtype == 'bool'):
+            new_indices = logic_mask.nonzero()[0]
+        else:
+            new_indices = np.array(logic_mask)
+
+        if hasattr(self.mask, 'dtype') and (self.mask.dtype == 'bool'):
+            orig_indices = self.mask.nonzero()[0]
+        else:
+            orig_indices = np.array(self.mask)
+
+        self.mask[:] = False
+        and_indices = np.intersect1d(new_indices, orig_indices)
+        self.mask[and_indices.astype(np.uint64)] = True
+
     def mask_to_n_loudest_clustered_events(self, rank_method,
+                                           ranking_threshold=6,
                                            n_loudest=10,
                                            cluster_window=10):
         """Edits the mask property of the class to point to the N loudest
-        single detector events as ranked by ranking statistic. Events are
-        clustered so that no more than 1 event within +/- cluster-window will
-        be considered."""
+        single detector events as ranked by ranking statistic.
+
+        Events are clustered so that no more than 1 event within +/-
+        cluster_window will be considered. Can apply a threshold on the
+        ranking using ranking_threshold
+        """
 
-        # If this becomes memory intensive we can optimize
         sds = rank_method.single(self.trig_dict())
```

Comment:
Could/should these lines be computed using `trigs_f.select` as well? Would be good to avoid reading in all trigger's information if we don't need to in this method.